sync_diff_inspector/diff.go

```diff
@@ -14,652 +14,793 @@
 			}
+		}
+		isDataEqual, err := df.compareRows(ctx, info, dml)
+		if err != nil {
+			df.report.SetTableMeetError(schema, table, err)
+		}
+		isEqual = isEqual && isDataEqual
+	}
+	dml.node.State = state
+	id := rangeInfo.ChunkRange.Index
+	df.report.SetTableDataCheckResult(schema, table, isEqual, dml.rowAdd, dml.rowDelete, id)
+	return isEqual
+}
 
-			df.tables[schema][table] = &TableConfig{
-				TableInstance: TableInstance{
-					Schema: schema,
-					Table:  table,
-				},
-				TargetTableInfo: tableInfo,
-				Range:           "TRUE",
-				SourceTables:    sourceTables,
-			}
+func (df *Diff) BinGenerate(ctx context.Context, targetSource source.Source, tableRange *splitter.RangeInfo, count int64) (*splitter.RangeInfo, error) {
+	if count <= splitter.SplitThreshold {
+		return tableRange, nil
+	}
+	tableDiff := targetSource.GetTables()[tableRange.GetTableIndex()]
+	indices := dbutil.FindAllIndex(tableDiff.Info)
+	// if no index, do not split
+	if len(indices) == 0 {
+		log.Warn("cannot found an index to split and disable the BinGenerate",
+			zap.String("table", dbutil.TableName(tableDiff.Schema, tableDiff.Table)))
+		return tableRange, nil
+	}
+	var index *model.IndexInfo
+	// using the index
+	for _, i := range indices {
+		if tableRange.IndexID == i.ID {
+			index = i
+			break
 		}
 	}
+	// TODO use selectivity from utils.GetBetterIndex
+	// only support PK/UK
+	if !(index.Primary || index.Unique) {
+		log.Warn("BinGenerate only support PK/UK")
+		return tableRange, nil
+	}
+	if index == nil {
+		log.Warn("have indexs but cannot found a proper index to split and disable the BinGenerate",
+			zap.String("table", dbutil.TableName(tableDiff.Schema, tableDiff.Table)))
+		return tableRange, nil
+	}
+	log.Debug("index for BinGenerate", zap.String("index", index.Name.O))
+	indexColumns := utils.GetColumnsFromIndex(index, tableDiff.Info)
+	if len(indexColumns) == 0 {
+		log.Warn("fail to get columns of the selected index, directly return the origin chunk")
+		return tableRange, nil
+	}
 
-	return nil
+	return df.binSearch(ctx, targetSource, tableRange, count, tableDiff, indexColumns)
 }
 
-// AdjustTableConfig adjusts the table's config by check-tables and table-config.
-func (df *Diff) AdjustTableConfig(cfg *Config) (err error) {
-	if len(df.subTaskCfgs) != 0 {
-		return df.adjustTableConfigBySubTask(cfg)
+func (df *Diff) binSearch(ctx context.Context, targetSource source.Source, tableRange *splitter.RangeInfo, count int64, tableDiff *common.TableDiff, indexColumns []*model.ColumnInfo) (*splitter.RangeInfo, error) {
+	if count <= splitter.SplitThreshold {
+		return tableRange, nil
+	}
+	var (
+		isEqual1, isEqual2 bool
+		count1, count2     int64
+	)
+	tableRange1 := tableRange.Copy()
+	tableRange2 := tableRange.Copy()
+
+	chunkLimits, args := tableRange.ChunkRange.ToString(tableDiff.Collation)
+	limitRange := fmt.Sprintf("(%s) AND %s", chunkLimits, tableDiff.Range)
```

Comment:
See dumpling bug: https://github.com/pingcap/dumpling/issues/371#issuecomment-942106862
Please make sure other places won't have the similar problem.
```suggestion
	limitRange := fmt.Sprintf("(%s) AND (%s)", chunkLimits, tableDiff.Range)
```

Reply:
Already fix all the places where this bug occurs.